pgx/src/datum/time_stamp_with_timezone.rs

```diff
@@ -8,143 +8,169 @@ Use of this source code is governed by the MIT license that can be found in the
-                tm_yday: 0,
-                tm_isdst: 0,
-                tm_gmtoff: 0,
-                tm_zone: std::ptr::null_mut(),
-            };
-            let mut tz = 0i32;
-            let mut fsec = 0 as pg_sys::fsec_t;
-            let mut tzn = std::ptr::null::<std::os::raw::c_char>();
-            pg_sys::timestamp2tm(
-                datum.value() as i64,
-                &mut tz,
-                &mut tm,
-                &mut fsec,
-                &mut tzn,
-                std::ptr::null_mut(),
-            );
-            let date = time::Date::from_calendar_date(
-                tm.tm_year,
-                time::Month::try_from(tm.tm_mon as u8)
-                    .expect("Got month outside of range in TimestampWithTimeZone::from_datum"),
-                tm.tm_mday as u8,
-            )
-            .expect("failed to create date from TimestampWithTimeZone");
+    #[deprecated(
+        since = "0.5.0",
+        note = "the repr of pgx::TimestampWithTimeZone is no longer time::OffsetDateTime \
+    and this fn will be removed in a future version"
+    )]
+    pub fn new(time: time::PrimitiveDateTime, at_tz_offset: UtcOffset) -> Self {
+        let offset = time.assume_utc()
+                        .to_offset(
+                            UtcOffset::from_whole_seconds(-at_tz_offset.whole_seconds())
+                                .expect("Unexpected error in `UtcOffset::from_whole_seconds` during `TimestampWithTimeZone::new`")
+                        );
+        offset
+            .try_into()
+            .expect("unable to convert time::PrimitiveDateTime to pgx::TimestampWithTimeZone")
+    }
+}
 
-            let time = time::Time::from_hms_micro(
-                tm.tm_hour as u8,
-                tm.tm_min as u8,
-                tm.tm_sec as u8,
-                fsec as u32,
-            )
-            .expect("failed to create time from TimestampWithTimeZonez");
-
-            Some(TimestampWithTimeZone(
-                time::PrimitiveDateTime::new(date, time)
-                    .assume_utc()
-                    .to_offset(
-                        UtcOffset::from_whole_seconds(tz)
-                            .expect("Unexpected error in `UtcOffset::from_whole_seconds` during `TimestampWithTimeZone::from_datum`")
-                    ),
-            ))
+impl From<TimestampWithTimeZone> for i64 {
+    fn from(tstz: TimestampWithTimeZone) -> Self {
+        tstz.0
+    }
+}
+
+impl TryFrom<pg_sys::TimestampTz> for TimestampWithTimeZone {
+    type Error = TimestampConversionError;
+
+    fn try_from(value: pg_sys::TimestampTz) -> Result<Self, Self::Error> {
+        let usec = value as i64;
+        match usec {
+            i64::MIN => Ok(Self::NEG_INFINITY),
+            i64::MAX => Ok(Self::INFINITY),
+            // https://github.com/rust-lang/rust/issues/37854 for funny exlusive range guard
+            MIN_TIMESTAMP_USEC..=END_TIMESTAMP_USEC if usec < END_TIMESTAMP_USEC => {
```

Comment:
This is... not idiomatic. Please decrease the end constant by 1 and omit the guard.

Reply:
done